### ISSUE TITLE
v3.6.5: Juice#test_time_neg test failing on several  (32bit) Debian architectures 

### ISSUE THREAD

**Issue by boutil** [ohler55/oj#488]
Hi,

when packaging version 3.6.5 for Debian, I noticed that altough all tests pass on my x86_64 machine, builds for armel, armhf, i386, mips, mipsel, hppa, hurdi386 and powerpc fail, because of the following test failure:
```

  1) Error:
Juice#test_time_neg:
ArgumentError: time out of system range
    /<<PKGBUILDDIR>>/test/test_various.rb:412:in `dump'
    /<<PKGBUILDDIR>>/test/test_various.rb:412:in `test_time_neg'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:107:in `block (3 levels) in run'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:204:in `capture_exceptions'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:104:in `block (2 levels) in run'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:255:in `time_it'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:103:in `block in run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:350:in `on_signal'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:275:in `with_info_handler'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:102:in `run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:839:in `run_one_method'
    /usr/lib/ruby/vendor_ruby/minitest.rb:324:in `run_one_method'
    /usr/lib/ruby/vendor_ruby/minitest.rb:311:in `block (2 levels) in run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:310:in `each'
    /usr/lib/ruby/vendor_ruby/minitest.rb:310:in `block in run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:350:in `on_signal'
    /usr/lib/ruby/vendor_ruby/minitest.rb:337:in `with_info_handler'
    /usr/lib/ruby/vendor_ruby/minitest.rb:309:in `run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:159:in `block in __run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:159:in `map'
    /usr/lib/ruby/vendor_ruby/minitest.rb:159:in `__run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:136:in `run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:63:in `block in autorun'
```
See https://buildd.debian.org/status/package.php?p=ruby-oj for the build results and the logs.

I am able to do further tests under your guidance if needed.

Best wishes,

Cédric

**Comment by ohler55** [comment]
I might need some help. I'll start a branch for this tonight.

**Comment by ohler55** [comment]
I think the first step is to try some times and see what happens. Can you try this command in an irb?
```
require 'time'
secs = Time.parse("1900-01-01 00:18:59 UTC").tv_sec
```
What value is returned?

Now reverse the process.
```
t2 = Time.at(secs)
```

**Comment by boutil** [comment]
I did the test on an armhf machine. I got `-2208987661` for `sec` and `1900-01-01 00:18:59 +0000` for `t2`.
These are the same results as on my amd64 local machine. It is for both ruby 2.5.1p57.

**Comment by ohler55** [comment]
That's what I would have expected for no errors. Darn. I checked in a branch that turns on some tracing for the test. Maybe that will isolate the issue as either dumping or loading. The branch is `debian`.

**Comment by boutil** [comment]
Here is the output on armhf with the failing test:
```
# Running:

...............................................................................................................................................................*** time is 1900-01-01 00:18:59UTC with secsonds of -2208987661
E.....................................................................................................................................................{}
.............................................................................................................................................

Finished in 0.348914s, 1289.7147 runs/s, 2418.9315 assertions/s.

  1) Error:
Juice#test_time_neg:
ArgumentError: time out of system range
    /home/boutil/oj/test/test_various.rb:413:in `dump'
    /home/boutil/oj/test/test_various.rb:413:in `test_time_neg'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:107:in `block (3 levels) in run'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:204:in `capture_exceptions'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:104:in `block (2 levels) in run'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:255:in `time_it'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:103:in `block in run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:350:in `on_signal'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:275:in `with_info_handler'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:102:in `run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:839:in `run_one_method'
    /usr/lib/ruby/vendor_ruby/minitest.rb:324:in `run_one_method'
    /usr/lib/ruby/vendor_ruby/minitest.rb:311:in `block (2 levels) in run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:310:in `each'
    /usr/lib/ruby/vendor_ruby/minitest.rb:310:in `block in run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:350:in `on_signal'
    /usr/lib/ruby/vendor_ruby/minitest.rb:337:in `with_info_handler'
    /usr/lib/ruby/vendor_ruby/minitest.rb:309:in `run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:159:in `block in __run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:159:in `map'
    /usr/lib/ruby/vendor_ruby/minitest.rb:159:in `__run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:136:in `run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:63:in `block in autorun'

450 runs, 844 assertions, 0 failures, 1 errors, 0 skips
```

**Comment by boutil** [comment]
For comparison, the corresponding output on the amd64 machine
```
# Running:

..................................................................................{}
...............................................................*** time is 1900-01-01 00:18:59 UTC with secsonds of -2208987661
*** json is -2208987661.000000000
.................................................................................................................................................................................................................................................................................................................

Finished in 0.050759s, 8865.4631 runs/s, 16647.3695 assertions/s.

450 runs, 845 assertions, 0 failures, 0 errors, 0 skips
```

**Comment by ohler55** [comment]
Added some debug prints to the time dump in C.

**Comment by boutil** [comment]
Here is the relevant piece of output on armhf:
```
-------- running test/tests.rb test -------
Run options: --seed 22499

# Running:

......................................................................................................................................................................................................................................................*** time is 1900-01-01 00:18:59 UTC with secsonds of -2208987661
E.............................................{}
..............*** nsec: 158098812
*** nsec: 123456000
*** nsec: 123456788
*** nsec: 123456788
*** nsec: 123456788
*** nsec: 123456788
....................................................................................................................*** nsec: 158098812
*** nsec: 123456000
*** nsec: 123456788
*** nsec: 123456788
*** nsec: 123456788
*** nsec: 123456788
............................

Finished in 0.344978s, 1304.4315 runs/s, 2446.5337 assertions/s.

  1) Error:
Juice#test_time_neg:
ArgumentError: time out of system range
    /home/boutil/oj/test/test_various.rb:413:in `dump'
    /home/boutil/oj/test/test_various.rb:413:in `test_time_neg'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:107:in `block (3 levels) in run'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:204:in `capture_exceptions'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:104:in `block (2 levels) in run'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:255:in `time_it'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:103:in `block in run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:350:in `on_signal'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:275:in `with_info_handler'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:102:in `run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:839:in `run_one_method'
    /usr/lib/ruby/vendor_ruby/minitest.rb:324:in `run_one_method'
    /usr/lib/ruby/vendor_ruby/minitest.rb:311:in `block (2 levels) in run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:310:in `each'
    /usr/lib/ruby/vendor_ruby/minitest.rb:310:in `block in run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:350:in `on_signal'
    /usr/lib/ruby/vendor_ruby/minitest.rb:337:in `with_info_handler'
    /usr/lib/ruby/vendor_ruby/minitest.rb:309:in `run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:159:in `block in __run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:159:in `map'
    /usr/lib/ruby/vendor_ruby/minitest.rb:159:in `__run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:136:in `run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:63:in `block in autorun'

450 runs, 844 assertions, 0 failures, 1 errors, 0 skips
*** nsec: 158098812
*** nsec: 123456000
*** nsec: 123456788
*** nsec: 123456788
*** nsec: 123456788
*** nsec: 123456788
*** nsec: 405498874
*** nsec: 405498874
E: Test test/tests.rb has failed
```

**Comment by ohler55** [comment]
One more try for tonight. Pull then just run the one test.
```
test_various.rb -n test_time_neg
```

**Comment by boutil** [comment]
Here is the output for this test on armhf with your last commit:
```
Run options: -n test_time_neg --seed 14134

# Running:


**** test_time_neg
*** time is 1900-01-01 00:18:59 UTC with secsonds of -2208987661
E

Finished in 0.008309s, 120.3563 runs/s, 0.0000 assertions/s.

  1) Error:
Juice#test_time_neg:
ArgumentError: time out of system range
    test/test_various.rb:414:in `dump'
    test/test_various.rb:414:in `test_time_neg'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:107:in `block (3 levels) in run'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:204:in `capture_exceptions'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:104:in `block (2 levels) in run'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:255:in `time_it'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:103:in `block in run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:350:in `on_signal'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:275:in `with_info_handler'
    /usr/lib/ruby/vendor_ruby/minitest/test.rb:102:in `run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:839:in `run_one_method'
    /usr/lib/ruby/vendor_ruby/minitest.rb:324:in `run_one_method'
    /usr/lib/ruby/vendor_ruby/minitest.rb:311:in `block (2 levels) in run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:310:in `each'
    /usr/lib/ruby/vendor_ruby/minitest.rb:310:in `block in run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:350:in `on_signal'
    /usr/lib/ruby/vendor_ruby/minitest.rb:337:in `with_info_handler'
    /usr/lib/ruby/vendor_ruby/minitest.rb:309:in `run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:159:in `block in __run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:159:in `map'
    /usr/lib/ruby/vendor_ruby/minitest.rb:159:in `__run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:136:in `run'
    /usr/lib/ruby/vendor_ruby/minitest.rb:63:in `block in autorun'

1 runs, 0 assertions, 0 failures, 1 errors, 0 skips
```

**Comment by ohler55** [comment]
Zeroing in on it. I added some more debugging. I think I have the line where the issue is or rather one of two lines now.

**Comment by ohler55** [comment]
Ok, now I'm confused. Different paths on each run. Back to the drawing board.

**Comment by ohler55** [comment]
Can you run `ruby -e "puts RUBY_PLATFORM"` on each system you are using. I'll add something to the `extconf.rb` to not use `rb_time_timespec` since that seems to be the call that is having issues.

**Comment by boutil** [comment]
Sorry for the delay.
Here is the output of RUBY_PLATFORM (first column) for the failing officially Debian supported architectures (second column):
```
arm-linux-gnueabi	armel
arm-linux-gnueabihf	armhf
i386-linux-gnu	i386
mips-linux-gnu	mips
mipsel-linux-gnu	mipsel
```

**Comment by boutil** [comment]
There is also
`hppa-linux-gnu` (hppa)
powerpc should be `powerpc-linux-gnu`
and hurd is `i386-gnu` .

**Comment by ohler55** [comment]
Pushed the change. Lets see if it works.

**Comment by boutil** [comment]
no luck
with the new extconf:
```
Run options: -n test_time_neg --seed 64663

# Running:


**** test_time_neg
*** time is 1900-01-01 00:18:59 UTC with secsonds of -2208987661
*** does not have timespec
*** rsec - class Integer for -2208987661
*** has nano time
*** nsec: 0
*** json is 2085979635.000000000
F

Finished in 0.016165s, 61.8603 runs/s, 61.8603 assertions/s.

  1) Failure:
Juice#test_time_neg [test/test_various.rb:416]:
--- expected
+++ actual
@@ -1,2 +1,2 @@
 # encoding: UTF-8
-"-2208987661.000000000"
+"2085979635.000000000"


1 runs, 1 assertions, 1 failures, 0 errors, 0 skips
```

**Comment by ohler55** [comment]
It is an improvement. No crash. Now to figure out why the numbers are off. I’m guessing it is a 64 vs 32 bit issue. I’ll put together another set of changes to nail it down.

**Comment by ohler55** [comment]
It does seem to be a 32 vs 64 bit issue as -2208987661 as an unsigned 32 bit is 2085979635. I made a change that may take care of that. Can you try again?

**Comment by boutil** [comment]
All tests pass on arm and mipsel architectures \o/

**Comment by ohler55** [comment]
Cool, that must have been it. I'll clean up and release. Thanks.

**Comment by ohler55** [comment]
Released 3.6.6 with the changes.

**Comment by boutil** [comment]
The 3.6.6 built fine on all Debian official architectures
https://buildd.debian.org/status/package.php?p=ruby-oj
thanks!
